File: src/llm_vs_zombies/app_update_anchor.py
"""Explicit prewarm App update-count initialization; never normalizes state."""
from __future__ import annotations
import copy

from . import sound_effects

MODE = "initial_app_update_anchor_v1"
METHOD = "app_update_anchor"
EVENT = "app_update_anchored"
MAX_VALUE = 0x7fffffff
SPEC = {"mode": MODE, "available": True, "field": "LawnApp+0x484/mUpdateCount",
    "phase": "after_seed_before_warm", "requires_audio_mode": sound_effects.MODE,
    "once_per_epoch": True, "value_range": [0, MAX_VALUE],
    "demo_policy": "reject_original_demo_recording_or_playback",
    "original_engine_bitwise_unmodified": False, "live_verified": False, "receipt_event": EVENT}


def fail(message):
    from .audit_compare import EvidenceError
    raise EvidenceError("App update anchor: " + message)


def target(value):
    if type(value) is not int or not 0 <= value <= MAX_VALUE:
        fail("target must be an integer in 0..2147483647")
    return value


def mode(game):
    if "app_update_anchor" not in game:
        return None
    if not sound_effects.same(game["app_update_anchor"], SPEC):
        fail("unsupported or incomplete declared configuration")
    if sound_effects.mode(game) != sound_effects.MODE:
        fail("initial anchoring requires the explicit allocation-none audio mode")
    from .audit_compare import draw_mode, DRAW_SCHEDULE_MODE
    if draw_mode(game) != DRAW_SCHEDULE_MODE:
        fail("initial anchoring requires controlled warm drawing")
    return MODE


def negotiate(hello):
    declared = mode(hello.get("game", {}))
    caps = hello.get("capabilities", {})
    for key in (MODE, METHOD):
        actual = caps.get(key, False)
        if type(actual) is not bool or actual != bool(declared):
            fail("hello capability/configuration mismatch")
    return declared


def target_from_recipe(recipe):
    if METHOD not in recipe:
        return None
    block = recipe[METHOD]
    if (not isinstance(block, dict) or set(block) != {"configuration", "app_update_count"}
            or not sound_effects.same(block["configuration"], SPEC)):
        fail("initialization recipe configuration is invalid")
    return target(block["app_update_count"])


def receipt(value, game, *, requested=None, before_version=None, after_version=None, seed=None):
    from .audit_compare import SCHEMA, _seeded_rng, digests, version
    if not mode(game):
        fail("receipt lacks an explicit initialization contract")
    keys = {"schema", "mode", "before", "after", "requested", "before_state", "after_state", "demo_before", "demo_after",
            "before_version", "after_version"}
    if (not isinstance(value, dict) or set(value) != keys or value["schema"] != "lvz.app-update-anchor.v1"
            or value["mode"] != MODE or not sound_effects.uint(value["before"])):
        fail("receipt shape or original counter is invalid")
    target(value["before"])
    demo = value["demo_before"]
    if (not isinstance(demo, dict) or set(demo) != {"00000510", "00000511", "00000578", "0000049c", "000004a0"}
            or any(not sound_effects.uint(n) for n in demo.values()) or demo["00000510"] != 0 or demo["00000511"] != 0
            or demo["000004a0"] > 255
            or not sound_effects.same(demo, value["demo_after"])):
        fail("original demo is active or its raw guard fields changed")
    wanted = target(value["requested"])
    if type(value["after"]) is not int or value["after"] != wanted:
        fail("actual readback differs from requested target")
    if requested is not None and wanted != target(requested):
        fail("receipt belongs to a different requested target")
    pre, post = version(value["before_version"]), version(value["after_version"])
    if pre["tick"] != 0 or post != dict(pre, revision=pre["revision"] + 1):
        fail("anchor must be one initialization revision at tick zero")
    if before_version is not None and not sound_effects.same(pre, before_version):
        fail("receipt before version differs from actual request boundary")
    if after_version is not None and not sound_effects.same(post, after_version):
        fail("receipt after version differs from actual reply/event boundary")
    for key, counter in (("before_state", value["before"]), ("after_state", wanted)):
        state = value[key]
        if (not isinstance(state, dict) or state.get("schema") != SCHEMA
                or state.get("rng", {}).get("target") != game.get("target")
                or state.get("app", {}).get("ui") != 3
                or not sound_effects.same(state.get("draw_schedule"),
                    {"mode": "deterministic_draw_schedule_v1", "warm_frames": 0, "step_frames": 0})):
            fail("receipt lacks full ready prewarm state")
        sound = sound_effects.state(state, game)
        if sound["app_update_count"] != counter:
            fail("receipt counter differs from captured state")
        digests(state)  # Retain ordinary integer-bit/schema validation.
        if seed is not None:
            if not sound_effects.uint(seed) or not sound_effects.same(state["rng"].get("instances"), _seeded_rng(seed)):
                fail("anchor did not preserve the actual freshly seeded RNG")
    expected = copy.deepcopy(value["before_state"])
    expected["sound_effects"]["app_update_count"] = wanted
    if not sound_effects.same(expected, value["after_state"]):
        fail("anchor changed captured state outside the declared App counter")
    return value


def initial(marker):
    declared = mode(marker["identity"]["game"])
    wanted = target_from_recipe(marker["initialization"])
    if not declared:
        if wanted is not None:
            fail("recipe cannot upgrade an older execution mode")
        return
    if wanted is None:
        fail("new mode requires an explicit initial anchor recipe")
    if marker["observation"].get("app_update_anchored") is not True:
        fail("B0 observation does not confirm the actual one-shot anchor")
    sound = sound_effects.state(marker["state"], marker["identity"]["game"])
    if sound["app_update_count"] != wanted:
        fail("B0 actual App counter differs from the recorded target")


def semantics(value, *, map_version=lambda value: value):
    """Compare effects; each raw before state remains independently validated."""
    if value is None:
        return None
    result = copy.deepcopy({k: v for k, v in value.items() if k not in {"before", "before_state"}})
    result["before_version"] = map_version(result["before_version"])
    result["after_version"] = map_version(result["after_version"])
    return result


class Evidence:
    def __init__(self, game):
        self.game, self.enabled = game, bool(mode(game))
        self.anchor = self.anchor_event = self.seed_event = None
        self.warm_started = self.seen_frame = False

    def event(self, event):
        kind = event["kind"]
        if kind == "app_update_anchor_failed":
            fail("native anchor failure invalidates strict evidence")
        if kind == EVENT:
            if not self.enabled or self.anchor is not None or self.warm_started or self.seen_frame:
                fail("undeclared, duplicate, or late initial anchor")
            payload = event["payload"]
            if (set(payload) != {"request_id", "anchor"} or not isinstance(payload["request_id"], str)
                    or not payload["request_id"] or self.seed_event is None):
                fail("anchor lacks its actual request or preceding seed event")
            seed = self.seed_event["payload"]["seed"]
            value = receipt(payload["anchor"], self.game, after_version=event["version"], seed=seed)
            seed_version = self.seed_event["version"]
            before = value["before_version"]
            if seed_version["epoch"] != before["epoch"] or seed_version["tick"] != 0 or seed_version["revision"] > before["revision"]:
                fail("seed does not precede the anchor in the same initialization epoch")
            self.anchor, self.anchor_event = copy.deepcopy(value), copy.deepcopy(event)
        elif self.enabled and kind == "rng_seeded":
            if self.anchor is not None or self.warm_started or self.seen_frame or not sound_effects.uint(event["payload"].get("seed")):
                fail("seed changed after anchoring or has invalid evidence")
            self.seed_event = copy.deepcopy(event)
        elif self.enabled and kind in {"rng_restored", "clocks_restored"}:
            if self.anchor is not None or self.warm_started or self.seen_frame:
                fail("initialization changed after the one-shot anchor")
            if kind == "rng_restored":
                self.seed_event = None
        elif self.enabled and kind == "render_preparing":
            if self.anchor is None or self.warm_started:
                fail("warm drawing requires exactly one completed anchor")
            expected = dict(self.anchor["after_version"], revision=self.anchor["after_version"]["revision"] + 1)
            if not sound_effects.same(expected, event["version"]):
                fail("warm preparation is not the next initialization revision")
            self.warm_started = True
        elif self.enabled and kind in {"request_started", "action"} and not self.warm_started:
            fail("simulation action precedes the required initial anchor/warm boundary")

    def frame(self, frame):
        if self.enabled and (self.anchor is None or not self.warm_started):
            fail("simulation frame lacks the required initial anchor")
        self.seen_frame = True

    def initial(self, marker):
        initial(marker)
        if not self.enabled:
            return
        wanted = target_from_recipe(marker["initialization"])
        if self.anchor is None or not self.warm_started or self.anchor["after"] != wanted:
            fail("B0 lacks matching actual anchored initialization evidence")
        expected = dict(self.anchor["after_version"], revision=self.anchor["after_version"]["revision"] + 1)
        if not sound_effects.same(expected, marker["observation"]["version"]):
            fail("B0 is not the actual next warm boundary")
        if marker["initialization"].get("seed") != self.seed_event["payload"]["seed"]:
            fail("recipe seed differs from the native seed before anchor")

File: determinism/app_update_anchor.cpp
#include "app_update_anchor.hpp"
#include "memory.hpp"
#include "launcher/silent_audio.hpp"
#include <limits>
namespace lvz::determinism {
using Json=nlohmann::json;
Json AppUpdateAnchorManifest(){return {{"mode",AppUpdateAnchorMode},{"available",true},
    {"field","LawnApp+0x484/mUpdateCount"},{"phase","after_seed_before_warm"},
    {"requires_audio_mode",lvz::silentaudio::Mode},{"once_per_epoch",true},
    {"value_range",{0,INT32_MAX}},{"original_engine_bitwise_unmodified",false},
    {"live_verified",false},{"receipt_event","app_update_anchored"},
    {"demo_policy","reject_original_demo_recording_or_playback"}};}
namespace {
bool U32(const Json& value){return value.is_number_integer()&&(value.is_number_unsigned()?value.get<uint64_t>()<=UINT32_MAX
    :value.get<int64_t>()>=0&&value.get<int64_t>()<=UINT32_MAX);}
void NoDemo(const Json& demo){
    if(!demo.is_object()||demo.size()!=5)throw std::runtime_error("App anchor missing demo guard evidence");
    for(auto key:{"00000510","00000511","00000578","0000049c","000004a0"})
        if(!U32(demo.at(key)))throw std::runtime_error("App anchor invalid actual demo field");
    if(demo.at("000004a0").get<uint32_t>()>255)throw std::runtime_error("App anchor demo marker is not a byte");
    if(demo.at("00000510")!=0||demo.at("00000511")!=0)throw std::runtime_error("App anchor rejects original demo recording/playback");
}
void EmptyAudio(const Json& state){
    const auto& audio=state.at("sound_effects");
    if(audio.at("mode")!=lvz::silentaudio::Mode||audio.at("slots_empty")!=true||audio.at("patch_owned")!=true
        ||audio.at("errors")!=0||!U32(audio.at("calls")))throw std::runtime_error("App anchor requires healthy allocation-none audio");
    const auto& histories=audio.at("histories");const auto& channels=audio.at("channels");
    if(!histories.is_array()||histories.size()!=110||!channels.is_array()||channels.size()!=32)
        throw std::runtime_error("App anchor requires actual complete Foley/channel evidence");
    for(const auto& history:histories){const auto& slots=history.at("slots");
        if(!slots.is_array()||slots.size()!=8)throw std::runtime_error("App anchor missing Foley slots");
        for(const auto& slot:slots)if(!slot.is_array()||slot.size()!=5||!U32(slot[0])||!U32(slot[1])||slot[0]!=0||slot[1]!=0)
            throw std::runtime_error("App anchor rejects nonempty Foley instance/refcount");}
    for(const auto& channel:channels)if(!U32(channel)||channel!=0)throw std::runtime_error("App anchor rejects nonempty sound channel");
}
void Seeded(const Json& state,uint32_t seed){
    const auto& rng=state.at("rng").at("instances");const auto& mt=rng.at("global_mt");
    const auto& words=mt.at("words");
    if(mt.at("cursor")!=624||!words.is_array()||words.size()!=624||rng.at("game_thread_crt").at("state")!=seed)
        throw std::runtime_error("App anchor RNG readback differs from explicit seed");
    uint32_t word=seed?seed:4357u;
    for(unsigned i=0;i<624;++i){if(words[i]!=word)throw std::runtime_error("App anchor MT word differs from explicit seed");word=1812433253u*(word^(word>>30))+i+1;}
}
}
Json InitialAppUpdateAnchor::Apply(uintptr_t address,uint32_t requested,uint32_t seed,bool audioEnabled,
                                  bool seeded,bool warmed,const std::function<Json()>& capture,const std::function<Json()>& captureDemo){
    Json receipt={{"schema","lvz.app-update-anchor.v1"},{"mode",AppUpdateAnchorMode},{"requested",requested},
        {"before",nullptr},{"after",nullptr},{"before_state",nullptr},{"after_state",nullptr},
        {"demo_before",nullptr},{"demo_after",nullptr}};
    bool wrote=false;
    try {
        if(!audioEnabled||!seeded||warmed||attempted_||requested>INT32_MAX)
            throw std::runtime_error("App anchor requires allocation-none, explicit seed, no warm, and an unused initial anchor");
        const auto demo=captureDemo();receipt["demo_before"]=demo;NoDemo(demo);
        auto before=capture();receipt["before_state"]=before;
        EmptyAudio(before);Seeded(before,seed);
        const auto& count=before.at("sound_effects").at("app_update_count");
        if(!U32(count)||count.get<uint32_t>()>INT32_MAX)throw std::runtime_error("App anchor original counter is outside signed range");
        receipt["before"]=count;
        if(!Accessible(address,4,true)||Read<uint32_t>(address)!=count.get<uint32_t>())
            throw std::runtime_error("App anchor writable target differs from actual snapshot");
        attempted_=true;wrote=true;
        // One explicit original field write; even a source no-op is a consumed
        // anchor. A post-write verification failure is never rolled back.
        std::memcpy(reinterpret_cast<void*>(address),&requested,4);
        auto after=capture();receipt["after_state"]=after;
        receipt["after"]=after.at("sound_effects").at("app_update_count");
        receipt["demo_after"]=captureDemo();
        if(receipt["demo_after"]!=demo)throw std::runtime_error("App anchor changed original demo state");
        auto expected=std::move(before);expected["sound_effects"]["app_update_count"]=requested;
        if(after!=expected||Read<uint32_t>(address)!=requested)
            throw std::runtime_error("App anchor changed fields other than the declared actual counter");
        EmptyAudio(after);Seeded(after,seed);applied_=true;
        return {{"ok",true},{"write_attempted",true},{"anchor",std::move(receipt)}};
    }catch(const std::exception& error){return {{"ok",false},{"write_attempted",wrote},{"error",error.what()},{"anchor",std::move(receipt)}};}
}
}
